test/regress_ws.c

```diff
@@ -0,0 +1,362 @@
+
+	data = evbuffer_pullup(buf, data_len);
+
+	fin = !!(*data & 0x80);
+	opcode = *data & 0x0F;
+	mask = !!(*(data + 1) & 0x80);
+	payload_len = *(data + 1) & 0x7F;
+
+	header_len = 2 + (mask ? 4 : 0);
+
+	if (payload_len < 126) {
+		if (header_len > data_len)
+			return NULL;
+
+	} else if (payload_len == 126) {
+		header_len += 2;
+		if (header_len > data_len)
+			return NULL;
+
+		payload_len = ntohs(*(uint16_t *)(data + 2));
+
+	} else if (payload_len == 127) {
+		header_len += 8;
+		if (header_len > data_len)
+			return NULL;
+
+		payload_len = ntohll(*(uint64_t *)(data + 2));
+	}
+
+	if (header_len + payload_len > data_len)
+		return NULL;
+
+	mask_key = data + header_len - 4;
+	for (size_t i = 0; mask && i < payload_len; i++)
+		data[header_len + i] ^= mask_key[i % 4];
+
+	*out_len = payload_len;
+
+	/* text */
+	if (opcode == 0x01) {
+		out_buf = calloc(payload_len + 1, 1);
+	} else { /* binary */
+		out_buf = malloc(payload_len);
+	}
+	memcpy(out_buf, (const char *)data + header_len, payload_len);
+
+	if (!fin) {
+		*options = 1;
+	}
+
+	evbuffer_drain(buf, header_len + payload_len);
+	return out_buf;
+}
+
+static void
+send_ws_msg(struct evbuffer *buf, const char *msg)
+{
+	size_t len = strlen(msg);
+	uint8_t a = 0, b = 0, c = 0, d = 0;
+	uint8_t mask_key[4] = {1, 2, 3, 4}; /* should be random */
+	uint8_t m;
+
+	a |= 1 << 7; /* fin */
+	a |= 1;		 /* text frame */
+
+	b |= 1 << 7; /* mask */
+
+	/* payload len */
+	if (len < 126) {
+		b |= len;
+	} else if (len < (1 << 16)) {
+		b |= 126;
+		c = htons(len);
+	} else {
+		b |= 127;
+		d = htonll(len);
+	}
+
+	evbuffer_add(buf, &a, 1);
+	evbuffer_add(buf, &b, 1);
+
+	if (c)
+		evbuffer_add(buf, &c, sizeof(c));
+	else if (d)
+		evbuffer_add(buf, &d, sizeof(d));
+
+	evbuffer_add(buf, &mask_key, 4);
+
+	for (size_t i = 0; i < len; i++) {
+		m = msg[i] ^ mask_key[i % 4];
+		evbuffer_add(buf, &m, 1);
+	}
+}
+
+static void
+http_ws_readcb_phase2(struct bufferevent *bev, void *arg)
+{
+	struct evbuffer *input = bufferevent_get_input(bev);
+	struct evbuffer *output = bufferevent_get_output(bev);
+
+	while (evbuffer_get_length(input) >= 2) {
+		size_t len = 0;
+		unsigned options = 0;
+		char *msg;
+
+		msg = receive_ws_msg(input, &len, &options);
+		if (msg) {
+			if (!strcmp(msg, "Server: hello")) {
+				send_ws_msg(output, "Send echo");
+				test_ok++;
+			} else if (!strcmp(msg, "Reply echo")) {
+				send_ws_msg(output, "Close");
+				test_ok++;
+			} else {
+				test_ok--;
+			}
+			free(msg);
+		}
+	}
+}
+
+static void
+http_ws_readcb_hdr(struct bufferevent *bev, void *arg)
+{
+	struct evbuffer *input = bufferevent_get_input(bev);
+	struct evbuffer *output = bufferevent_get_output(bev);
+	size_t nread = 0, n = 0;
+	char *line;
+
+	while ((line = evbuffer_readln(input, &nread, EVBUFFER_EOL_CRLF))) {
+		if (n == 0 &&
+			!strncmp(line, "HTTP/1.1 101 ", strlen("HTTP/1.1 101 "))) {
+			test_ok++;
+		} else if (!strcmp(line,
+					   "Sec-WebSocket-Accept: HSmrc0sMlYUkAGmm5OPpG2HaGWk=")) {
```

Comment:
It is fine to use constant here, but using decode/encode here will be better.

Reply:
I used these "well-known" WS protocol handshake from https://en.wikipedia.org/wiki/WebSocket#Protocol_handshake and thought it would be better to have them in strings for tests to "prove" protocol correctness

Reply:
Good explanation, agree.